storage/pebble/pebble.go

```
@@ -0,0 +1,202 @@
+	return pbdb, nil
+}
+
+func (pdb *PebbleDB) Close() error {
+	return pdb.DB.Close()
+}
+
+func (pdb *PebbleDB) Get(prefix []byte, key []byte) (basedb.Obj, bool, error) {
+	return getter(key, func(key []byte) ([]byte, io.Closer, error) {
+		return pdb.DB.Get(append(prefix, key...))
+	})
+}
+
+func (pdb *PebbleDB) Set(prefix, key, value []byte) error {
+	return pdb.DB.Set(append(prefix, key...), value, pebble.Sync)
+}
+
+func (pdb *PebbleDB) Delete(prefix, key []byte) error {
+	return pdb.DB.Delete(append(prefix, key...), pebble.Sync)
+}
+
+func (pdb *PebbleDB) GetMany(prefix []byte, keys [][]byte, fn func(basedb.Obj) error) error {
+	return manyGetter(pdb.logger, keys, func(key []byte) ([]byte, io.Closer, error) {
+		fullKey := append(prefix, key...)
+		return pdb.DB.Get(fullKey)
+	}, fn)
+}
+
+func makePrefixIter(dbOrBatch pebble.Reader, prefix []byte) (*pebble.Iterator, error) {
+	keyUpperBound := func(b []byte) []byte {
+		end := make([]byte, len(b))
+		copy(end, b)
+		for i := len(end) - 1; i >= 0; i-- {
+			end[i] = end[i] + 1
+			if end[i] != 0 {
+				return end[:i+1]
+			}
+		}
+		return nil // no upper-bound
+	}
+
+	prefixIterOptions := func(prefix []byte) *pebble.IterOptions {
+		return &pebble.IterOptions{
+			LowerBound: prefix,
+			UpperBound: keyUpperBound(prefix),
+		}
+	}
+
+	return dbOrBatch.NewIter(prefixIterOptions(prefix))
+}
+
+func (pdb *PebbleDB) GetAll(prefix []byte, fn func(int, basedb.Obj) error) (err error) {
+	iter, err := makePrefixIter(pdb.DB, prefix)
+	if err != nil {
+		return err
+	}
+
+	defer func() { _ = iter.Close() }()
+
+	return allGetter(pdb.logger, iter, prefix, fn)
+}
+
+func (pdb *PebbleDB) Begin() basedb.Txn {
+	txn := pdb.NewIndexedBatch()
+	return newPebbleTxn(pdb.logger, txn)
+}
+
+func (pdb *PebbleDB) Using(rw basedb.ReadWriter) basedb.ReadWriter {
+	if rw == nil {
+		return pdb
+	}
+	return rw
+}
+
+func (pdb *PebbleDB) UsingReader(r basedb.Reader) basedb.Reader {
+	if r == nil {
+		return pdb
+	}
+	return r
+}
```

Shouldn't we let it panic in case `basedb.Reader` is not provided while `UsingReader` call is made ? Sounds like a bug to me - so panic seems preferable.

Also applies to `Using` func above

---

these `Using...` methods were ported from the badger db implementation to maintain feature parity.

from what I see it's been implemented like this to allow for a certain flexibility - in parts of code where we fetch data sequentially through the same reader - we can pass the reader object - and in parts where we use this method as a callback we are safe to pass nil:

https://vscode.dev/github/ssvlabs/ssv/blob/feat/add-pebbledb-support/operator/validator/controller.go#L750

also usages of `getOperatorData`:

https://vscode.dev/github/ssvlabs/ssv/blob/feat/add-pebbledb-support/registry/storage/operators.go#L129

---

Right, I see, thanks for clarifying that, makes sense